### PR TITLE
[Model] TransactionRecord decomposeTransaction refactoring

### PR DESCRIPTION
**Pull request by furszy**
Initial step towards a cleaner and more understandable `TransactionRecord::decomposeTransaction` function.

No functional change. Have done mostly code refactoring with some cleanups and improvements.

### CODE REVIEW

**Comment by furszy** [comment]
Updated the PR with more changes. Couldn't resist myself.

The flow now should be much more "followable" and pleasant to read. Plus, it has some more performance improvements that have found on the way.

File: src/qt/transactionrecord.cpp [hunk 1]
```diff
@@ -373,13 +372,13 @@ void TransactionRecord::loadUnlockColdStake(const CWallet* wallet, const CWallet
...
     // Get the p2cs
-    const CScript* p2csScript = nullptr;
+    CScript p2csScript;
     bool isSpendable = false;
 
     for (const auto &input : wtx.vin) {
         const CWalletTx* tx = wallet->GetWalletTx(input.prevout.hash);
         if (tx && tx->vout[input.prevout.n].scriptPubKey.IsPayToColdStaking()) {
-            p2csScript = &tx->vout[input.prevout.n].scriptPubKey;
+            p2csScript = tx->vout[input.prevout.n].scriptPubKey;
```
Comment (random-zebra) [inline src/qt/transactionrecord.cpp:395]: (84a0785ddc9c887d159716339ac78d7d2d8380e1) why this change? using a pointer required one less copy of the script.

File: src/qt/transactionrecord.cpp [hunk 2]
```diff
@@ -312,56 +365,15 @@ QList<TransactionRecord> TransactionRecord::decomposeTransaction(const CWallet*
...
-                parts.append(sub);
-            }
-        } else {
-            //
-            // Mixed debit transaction, can't break down payees
-            //
-            parts.append(TransactionRecord(hash, nTime, wtx.GetTotalSize(), TransactionRecord::Other, "", nNet, 0));
-            parts.last().involvesWatchAddress = involvesWatchAddress;
+            decomposeDebitTransaction(wallet, wtx, nDebit, involvesWatchAddress, parts);
+            return parts;
```
Comment (random-zebra) [inline src/qt/transactionrecord.cpp]: (e6c637960177e601b9d67b707a90e7a6f68f254a) the return value of `decomposeDebitTransaction` is always true, and it never gets used. 
Same for `decomposeSendToSelfTransaction` (in 5e894df145fb16785c8f43342f88e2d53d24ecfd) and `decomposeCreditTransaction` (in 162efba46715dfdfc516351530c4fbda9ea842b1).

I think we should either keep (for consistency among all "decompose" functions)
```cpp
if (decomposeDebitTransaction(wallet, wtx, nDebit, involvesWatchAddress, parts))
    return parts;
```
or just make them `void` functions.

File: src/qt/transactionrecord.cpp [hunk 3]
```diff
@@ -246,11 +253,16 @@ bool TransactionRecord::decomposeDebitTransaction(const CWallet* wallet, const C
                                                   const CAmount& nDebit, bool involvesWatchAddress,
                                                   QList<TransactionRecord>& parts)
 {
+    // Return early if there are no outputs.
+    if (wtx.vout.empty()) {
+        return false;
+    }
+
     CAmount nTxFee = nDebit - wtx.GetValueOut();
+    TransactionRecord sub(wtx.GetHash(), wtx.GetTxTime(), wtx.GetTotalSize());
```
Comment (random-zebra) [inline src/qt/transactionrecord.cpp:262]: (cbf9c7c32cd8618d56194f7cb1bbda7f3a983deb) shouldn't we keep initializing a new sub-record for every output?
otherwise things like `address`, `debit`, `credit`, and `idx`, are not reset and might be carried over from a previous iteration.

File: src/qt/transactionrecord.cpp (hunk 1 above)
Comment (furszy) [reply src/qt/transactionrecord.cpp:395]: to suppress a small warning on the `ExtractAddress` that is below. If `p2csScript` pointer is null (which should never happen), then undefined behavior.

File: src/qt/transactionrecord.cpp (hunk 2 above)
Comment (furszy) [reply src/qt/transactionrecord.cpp]: yep, sounds good.

File: src/qt/transactionrecord.cpp (hunk 3 above)
Comment (furszy) [reply src/qt/transactionrecord.cpp:262]: good catch. Ideally, all of them are set on the loop cycle.
Going to cache the total tx size and generate a new `TransactionRecord` per cycle to be totally sure.

File: src/qt/transactionrecord.cpp (hunk 1 above)
Comment (random-zebra) [reply src/qt/transactionrecord.cpp:395]: Yes, I thought so. But we could keep the performance advantage of using  a pointer, simply adding a condition to the last line:
```cpp
if (p2csScript)
    ExtractAddress(*p2csScript, false, record.address);
```

Comment (furszy) [reply src/qt/transactionrecord.cpp:395]: done

**Comment by furszy** [comment]
Updated per review

File: src/qt/transactionrecord.cpp [hunk 4]
```diff
@@ -185,11 +185,18 @@ bool TransactionRecord::decomposeCreditTransaction(const CWallet* wallet, const
...
         isminetype mine = wallet->IsMine(txout);
         if (mine) {
-            TransactionRecord sub(hash, nTime, wtx.GetTotalSize());
+
+            // Cache total size to not be forced to recalculate it on every output
+            if (cachedTotalSize == 0) {
+                cachedTotalSize = wtx.GetTotalSize();
+            }
+
+            TransactionRecord sub(hash, nTime, cachedTotalSize);
```
Comment (random-zebra) [inline src/qt/transactionrecord.cpp]: We can simply initialize the tx size when we declare it, outside of the loop, same as we do below for  `decomposeDebitTransaction`.
```cpp
unsigned int txSize = wtx.GetTotalSize();
```

Comment (furszy) [reply src/qt/transactionrecord.cpp]: I did it in this way because it was calculating the size for nothing if there is no output from the wallet. 
But yeah.. thinking it more, `GetCredit` at the beginning of decomposeTransaction is being called with an `ISMINE_ALL` filter, which will prevent that from happening.

**Comment by furszy** [comment]
Updated per feedback. Added a new commit on top. Will squash them all once we complete the review process.
